libc/src/string/memory_utils/inline_bcmp.h

```diff
@@ -32,7 +32,8 @@
 
 namespace LIBC_NAMESPACE {
 
-LIBC_INLINE int inline_bcmp(const void *p1, const void *p2, size_t count) {
+__attribute__((flatten)) LIBC_INLINE int
```

Comment:
My understanding that flatten should be applied to top-level functions, rather than helper functions.
It's the top level function that needs to flattened. If overall amount of code is so large that compiler stops inlining, then it may as well flatten then the helper, but then emit a call to the helpers since the total amount of code is still large (above inlining threshold).
Middle helper functions may benefit from always_inline in some cases. But it has the same caveat -- if compiler will be forced to inline this function, then it may decide to unline something else instead. So always_inline will need to be applied all way down.

Reply:
AFAIU `flatten` is not recursive it only applies to the selected function.
Right now we have the function definition [1] (example with `memcmp`)
https://github.com/llvm/llvm-project/blob/8a233d8cfde4cd026b5dd12b56ea029749a02329/libc/src/string/memcmp.cpp#L16-L19
and the helper that delegates to real code [2]
https://github.com/llvm/llvm-project/blob/8a233d8cfde4cd026b5dd12b56ea029749a02329/libc/src/string/memory_utils/inline_memcmp.h#L36-L39

Out of the box, clang performs inlining of [2] in [1], but not the inlining of the [2]'s body leaving `memcmp` with 
```
00000000000c69c0 <memcmp>:
   c69c0: e9 1b 00 00 00                jmp     0xc69e0 <__llvm_libc::inline_memcmp_x86(__llvm_libc::cpp::byte const*, __llvm_libc::cpp::byte const*, unsigned long)>
...
```

Adding `[[gnu::flatten]]` to [1] makes no difference, adding `[[gnu::flatten]]` to [2] does inline the function in `memcmp`:
```
00000000000c6980 <memcmp>:
   c6980: 31 c0                         xorl    %eax, %eax
   c6982: 48 83 fa 08                   cmpq    $0x8, %rdx
   c6986: 0f 87 e4 00 00 00             ja      0xc6a70 <memcmp+0xf0>
   c698c: 48 8d 0d 9d e1 f6 ff          leaq    -0x91e63(%rip), %rcx
   c6993: 48 63 14 91                   movslq  (%rcx,%rdx,4), %rdx
...
```

Now I'm pretty sure that PGO will inline the call but there's no point in having an implementation of a memory function that only jumps to another function in the first place.

I can go with `[[always_inline]]` if you prefer although `[[gnu::flatten]]` has the added benefit that it will respect `[[noinline]]`.

Reply:
I see. Interesting, so this is not due to code size, but due to I guess compiler wanting to give different functions different addresses (or is inline_memcmp_x86 called from multiple places and compiler decides to reduce code size?).

The question is: can compiler tomorrow inline inline_memcmp_x86 into inline_memcmp (as we asked with flatten), but make memcmp jmp to inline_memcmp?
Another good question is how to test this. These things regress due to compiler bugs, or seemingly unrelated refactorings. Quality bugs are more frequent than crashing bugs because they are silent and go unnoticed.

Reply:
> inline_memcmp_x86 called from multiple places and compiler decides to reduce code size

`inline_memcmp` is called from within libc but it's also exposed as `memcmp`, that may be the reason.
Another reason (maybe?) is that `memcmp` is exported as a weak symbol so that the user can override it with another implementation if needed.

> The question is: can compiler tomorrow inline inline_memcmp_x86 into inline_memcmp (as we asked with flatten), but make memcmp jmp to inline_memcmp?
This is entirely possible, we should probably also mark them with `flatten` when they are trivial.

> Another good question is how to test this. These things regress due to compiler bugs, or seemingly unrelated refactorings. Quality bugs are more frequent than crashing bugs because they are silent and go unnoticed.

@nickdesaulniers for visibility.
Good question indeed. This could be part of the integration tests. We'd have to disassemble the function and make sure they do not consist in a single jump to another function.

Reply:
> > Another good question is how to test this. These things regress due to compiler bugs, or seemingly unrelated refactorings. Quality bugs are more frequent than crashing bugs because they are silent and go unnoticed.
> 
> @nickdesaulniers for visibility. Good question indeed. This could be part of the integration tests. We'd have to disassemble the function and make sure they do not consist in a single jump to another function.

FWIW we have such a test for ThreadSanitizer hot paths:
https://github.com/llvm/llvm-project/blob/main/compiler-rt/lib/tsan/check_analyze.sh
And I will send you a link to one internal test that does it.